### PR TITLE
[hail/ptypes] TableUnion, TableMultiWayZipJoin upcast

### PR DESCRIPTION
**Pull request by akotlar**

### CODE REVIEW

**Comment by akotlar** [comment]
Tim would you prefer RVD.unify to be separate from RVD.union, such that the caller controls upcast? Interface seems much simpler if RVD.union calls RVD.unify.

File: hail/src/main/scala/is/hail/rvd/RVD.scala [hunk 1]
```diff
@@ -1413,19 +1413,40 @@ object RVD {
...
+
   def union(
     rvds: Seq[RVD],
     joinKey: Int,
     ctx: ExecuteContext
   ): RVD = rvds match {
     case Seq(x) => x
     case first +: _ =>
-      assert(rvds.forall(_.rowPType == first.rowPType))
+      val newRVDs = unify(rvds)
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/rvd/RVD.scala]: union doesn't need to change actually. Keep the assertion, it's on the user (TableUnion) to make sure the ptypes are the same

File: hail/src/main/scala/is/hail/rvd/RVD.scala [hunk 2]
```diff
@@ -1413,19 +1413,40 @@ object RVD {
...
+  private def copyFromType(destPType: PType, srcPType: PType, srcRegionValue: RegionValue): RegionValue =
+    RegionValue(srcRegionValue.region, destPType.copyFromType(srcRegionValue.region, srcPType, srcRegionValue.offset, false))
+
+  def unify(
+    rvds: Seq[RVD]
+  ): IndexedSeq[RVD] = rvds match {
+    case rvds.length == 1 => rvds.toIndexedSeq
+    case _ =>
+      if(rvds.forall(_.rowPType == rvds.head.rowPType)) {
+        return rvds.toIndexedSeq
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/rvd/RVD.scala]: toFastIndexedSeq

File: hail/src/main/scala/is/hail/rvd/RVD.scala (hunk 1 above)
Comment (akotlar) [reply hail/src/main/scala/is/hail/rvd/RVD.scala]: sure thing

**Pull request merged**